File: src/modules/module_A/pageMetrics/titleExtractor.ts
import type { CheerioAPI } from 'cheerio';
import type { TitleData } from './types.js';

/**
 * Extract title information from a page
 */
export function extractTitle($: CheerioAPI): TitleData {
    const titleElement = $('title');
    const title = titleElement.text().trim();
    const titleLength = title.length;
    const hasMissingTitle = titleLength === 0;
    
    // TODO: Calculate pixel width based on font metrics
    // This would require font-specific calculations
    const titlePixelWidth = undefined;
    
    return {
        title: title || 'No title',
        titleLength,
        titlePixelWidth,
        hasMissingTitle
    };
}

/**
 * Validate title length (SEO best practices)
 */
export function validateTitleLength(titleLength: number): {
    isValid: boolean;
    message?: string;
} {
    if (titleLength === 0) {
        return { isValid: false, message: 'Missing title tag' };
    }
    
    if (titleLength < 30) {
        return { isValid: false, message: 'Title is too short (< 30 characters)' };
    }
    
    if (titleLength > 60) {
        return { isValid: false, message: 'Title is too long (> 60 characters)' };
    }
    
    return { isValid: true };
}

/**
 * Calculate approximate pixel width of title
 * Note: This is a rough estimation. Actual pixel width depends on font family, size, and weight
 */
export function estimateTitlePixelWidth(title: string): number {
    // Average character width in pixels (approximate for common fonts at 16px)
    const avgCharWidth = 8;
    return title.length * avgCharWidth;
}

File: npy-backend/modules/module_A/pagematrix/title_extractor.py
from bs4 import BeautifulSoup
import re
from modules.module_A.WebsiteCrawler.metrics.pixel_width import calculate_pixel_width as _pw


def _title_pixel_width(text: str) -> int:
    return _pw(text, font_size=20)

def extract_title(soup: BeautifulSoup) -> dict:
    """
    Extract title information from a page.
    """
    title_element = soup.find('title')
    title = title_element.get_text().strip() if title_element else ''
    title_length = len(title)
    has_missing_title = title_length == 0
    
    # Calculate pixel width
    title_pixel_width = _title_pixel_width(title)
    
    return {
        'title': title or 'No title',
        'titleLength': title_length,
        'titlePixelWidth': title_pixel_width,
        'hasMissingTitle': has_missing_title
    }

def validate_title_length(title_length: int) -> dict:
    """
    Validate title length (SEO best practices).
    """
    if title_length == 0:
        return {'isValid': False, 'message': 'Missing title tag'}
    
    if title_length < 30:
        return {'isValid': False, 'message': 'Title is too short (< 30 characters)'}
    
    if title_length > 60:
        return {'isValid': False, 'message': 'Title is too long (> 60 characters)'}
    
    return {'isValid': True}
